src/Ens.ts

```diff
@@ -0,0 +1,594 @@
+      return await this.getAddressForWrappedDomain(domain);
+    }
+
+    if (!resolver) {
+      const owner = await this.owner(domain);
+      if (isNullAddress(owner)) {
+        throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, {
+          domain,
+        });
+      }
+      throw new ResolutionError(ResolutionErrorCode.UnspecifiedResolver, {
+        domain,
+      });
+    }
+
+    const coinType = this.getCoinType(currencyTicker.toUpperCase());
+    return await this.fetchAddress(resolver, domain, coinType);
+  }
+
+  private async fetchAddress(
+    resolver: string,
+    domain: string,
+    coinType: string,
+  ): Promise<string | undefined> {
+    const formatsByCoinType = requireOrFail(
+      '@ensdomains/address-encoder',
+      '@ensdomains/address-encoder',
+      '>= 0.1.x <= 0.2.x',
+    ).formatsByCoinType;
+
+    const resolverContract = new EthereumContract(
+      resolverInterface(resolver, coinType),
+      resolver,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    const nodeHash = this.namehash(domain);
+    const addr: string =
+      coinType !== EthCoinIndex
+        ? await this.callMethod(resolverContract, 'addr', [nodeHash, coinType])
+        : await this.callMethod(resolverContract, 'addr', [nodeHash]);
+    if (isNullAddress(addr)) {
+      return undefined;
+    }
+
+    // eslint-disable-next-line no-undef
+    const data = Buffer.from(addr.replace('0x', ''), 'hex');
+    return formatsByCoinType[coinType].encoder(data);
+  }
+
+  private async getTextRecord(domain, key): Promise<string | undefined> {
+    const nodeHash = this.namehash(domain);
+    const resolver = await this.getResolverContract(domain);
+    return await this.callMethod(resolver, 'text', [nodeHash, key]);
+  }
+
+  private async getContentHash(domain: string): Promise<string | undefined> {
+    const contentHash = requireOrFail('content-hash', 'content-hash', '^2.5.2');
+    const nodeHash = this.namehash(domain);
+    const resolverContract = await this.getResolverContract(domain);
+    const contentHashEncoded = await this.callMethod(
+      resolverContract,
+      'contenthash',
+      [nodeHash],
+    );
+    const codec = contentHash.getCodec(contentHashEncoded);
+    if (codec !== 'ipfs-ns') {
+      return undefined;
+    }
+    return contentHash.decode(contentHashEncoded);
+  }
+
+  private async getResolverContract(
+    domain: string,
+    coinType?: string,
+  ): Promise<EthereumContract> {
+    const resolverAddress = await this.resolver(domain);
+    return new EthereumContract(
+      resolverInterface(resolverAddress, coinType),
+      resolverAddress,
+      this.provider,
+    );
+  }
+
+  private async callMethod(
+    contract: EthereumContract,
+    method: string,
+    params: (string | string[])[],
+  ): Promise<any> {
+    const result = await contract.call(method, params);
+    return result[0];
+  }
+
+  private checkSupportedDomain(domain: string): boolean {
```

Comment:
we should only support eth, luxe, xyz, kred, the rest maybe cause conflict if we support other naming services